### PR TITLE
250 McPAL frequency estimation and mean option for gamma in pwc 

### PR DESCRIPTION
**Pull request by AlexanderBenz**

### CODE REVIEW

File: skactiveml/classifier/_parzen_window_classifier.py [hunk 1]
```diff
@@ -139,6 +146,26 @@ def fit(self, X, y, sample_weight=None):
...
+        ]
+        if (
+            "gamma" in self.metric_dict_
+            and self.metric_dict["gamma"] == "mean"
+            and self.metric in allowed_mean_kernel_metrics
+        ):
+            delta = np.sqrt(2) * 1e-6
+            is_lbld = is_labeled(y, missing_label=self.missing_label)
+            N = np.max([2, np.sum(is_lbld)])
+            variance = np.var(X, axis=0) ** 2
```
Comment (dakot) [inline skactiveml/classifier/_parzen_window_classifier.py:165]:
```suggestion
            variance = np.var(X, axis=0)
```

File: skactiveml/classifier/_parzen_window_classifier.py [hunk 2]
```diff
@@ -139,6 +146,26 @@ def fit(self, X, y, sample_weight=None):
...
+            numerator = 2 * N * np.sum(variance)
+            denominator = (N - 1) * np.log((N - 1) / delta ** 2)
+            self.metric_dict_["gamma"] = np.sqrt(numerator / denominator)
```
Comment (dakot) [inline skactiveml/classifier/_parzen_window_classifier.py:168]:
```suggestion
            self.metric_dict_["gamma"] = 0.5 * denominator / numerator
```

File: skactiveml/classifier/_parzen_window_classifier.py [hunk 3]
```diff
@@ -139,6 +146,26 @@ def fit(self, X, y, sample_weight=None):
...
         )
+        allowed_mean_kernel_metrics = [
+            "rbf",
+            "chi2",
+            "polynomial",
+            "poly",
+            "laplacian",
+            "sigmoid",
+        ]
+        if (
```
Comment (dakot) [inline skactiveml/classifier/_parzen_window_classifier.py:157]: add manual test with values. change this section to a separate private function

File: skactiveml/pool/_probabilistic_al.py [hunk 4]
```diff
@@ -143,10 +165,40 @@ def query(
...
+            raise TypeError(
+                "clf has no predict_freq and metric was set to None"
+            )
+
         # Fit the classifier and predict frequencies.
         if fit_clf:
             clf = clone(clf).fit(X, y, sample_weight)
-        k_vec = clf.predict_freq(X_cand)
+        if self.metric is not None:
+            allowed_mean_kernel_metrics = [
```
Comment (dakot) [inline skactiveml/pool/_probabilistic_al.py:177]: get from PWC class

File: skactiveml/stream/_stream_probabilistic_al.py [hunk 5]
```diff
@@ -159,6 +162,19 @@ def query(
             return_utilities=return_utilities,
         )
         if self.metric is not None:
+            allowed_mean_kernel_metrics = [
```
Comment (dakot) [inline skactiveml/stream/_stream_probabilistic_al.py:165]: get from PWC class

**Review by dakot — CHANGES_REQUESTED** [review]

File: skactiveml/classifier/_parzen_window_classifier.py [hunk 6]
```diff
@@ -208,3 +237,10 @@ def predict_freq(self, X):
             for i in range(np.size(X, 0)):
                 F[i, :] = K[i, indices[i]] @ self.V_[indices[i], :]
         return F
+
+
+def _calculate_mean_gamma(N, variance, delta=(np.sqrt(2) * 1e-6)):
```
Comment (dakot) [inline skactiveml/classifier/_parzen_window_classifier.py:242]: @AlexanderBenz Please test this function independently from PWC. Use multiple values to check if it is correct.

File: skactiveml/classifier/_parzen_window_classifier.py [hunk 7]
```diff
@@ -139,6 +157,17 @@ def fit(self, X, y, sample_weight=None):
         self.metric_dict_ = (
             self.metric_dict if self.metric_dict is not None else {}
         )
+        if (
+            "gamma" in self.metric_dict_
+            and self.metric_dict["gamma"] == "mean"
+            and self.metric in ALLOWED_MEAN_KERNEL_METRICS
```
Comment (dakot) [inline skactiveml/classifier/_parzen_window_classifier.py:163]: gamma mean only for RBF kernel

File: skactiveml/pool/_probabilistic_al.py [hunk 8]
```diff
@@ -143,10 +165,32 @@ def query(
...
+            )
+
         # Fit the classifier and predict frequencies.
         if fit_clf:
             clf = clone(clf).fit(X, y, sample_weight)
-        k_vec = clf.predict_freq(X_cand)
+        if self.metric is not None:
+            if (
+                self.metric_dict is None
+                and self.metric in ALLOWED_MEAN_KERNEL_METRICS
```
Comment (dakot) [inline skactiveml/pool/_probabilistic_al.py:179]: mean only for RBF

File: skactiveml/stream/_stream_probabilistic_al.py [hunk 9]
```diff
@@ -55,8 +55,11 @@ class StreamProbabilisticAL(SingleAnnotatorStreamQueryStrategy):
         estimate for each class.
         If metric is set to None, the `predict_freq` function of the `clf` will
         be used instead. If this is not defined, an Exception is raised.
-    metric_dict : dict,
+    metric_dict : dict, default=None
         Any further parameters are passed directly to the kernel function.
+        If metric_dict is None and metric is 'rbf', 'chi2', 'polynomial',
+        'poly', 'laplacian', 'sigmoid', metric_dict is set to
+        {'gamma': 'mean'}.
```
Comment (dakot) [inline skactiveml/stream/_stream_probabilistic_al.py:62]: adapt: only for RBF

File: skactiveml/stream/_stream_probabilistic_al.py [hunk 10]
```diff
@@ -159,6 +162,11 @@ def query(
             return_utilities=return_utilities,
         )
         if self.metric is not None:
+            if (
+                self.metric_dict is None
+                and self.metric in ALLOWED_MEAN_KERNEL_METRICS
```
Comment (dakot) [inline skactiveml/stream/_stream_probabilistic_al.py:167]: only for rbf

**Review by dakot — CHANGES_REQUESTED** [review]

File: skactiveml/classifier/_parzen_window_classifier.py [hunk 11]
```diff
@@ -208,3 +226,9 @@ def predict_freq(self, X):
             for i in range(np.size(X, 0)):
                 F[i, :] = K[i, indices[i]] @ self.V_[indices[i], :]
         return F
+
+    def _calculate_mean_gamma(N, variance, delta=(np.sqrt(2) * 1e-6)):
+        numerator = 2 * N * np.sum(variance)
+        denominator = (N - 1) * np.log((N - 1) / delta ** 2)
+        gamma = 0.5 * denominator / numerator
```
Comment (mherde) [inline skactiveml/classifier/_parzen_window_classifier.py:233]: Check case when variance is zero and names of `numerator` and `denominator`.

**Review by mherde — CHANGES_REQUESTED** [review]

File: skactiveml/classifier/_parzen_window_classifier.py (hunk 11 above)
Comment (tpham93) [reply skactiveml/classifier/_parzen_window_classifier.py]: We switched the numerator and denominator name around.
Gamma was set to infinite when variance is 0. We added further handling to default to 1/n_features instead.

**Review by mherde — APPROVED** [review]

**Review by mherde — APPROVED** [review]

**Pull request merged**